File: src/codex.rs
use crate::{
    config::{Backend, Config, Route},
    process::{self, GroupChild},
    runner::{MAX_MESSAGE, Model},
    store::Store,
};
use anyhow::{Context, Result, bail, ensure};
use futures::StreamExt;
use serde_json::{Value, json};
use std::{collections::VecDeque, path::Path, process::Stdio, time::Duration};
use tokio::io::AsyncWriteExt;
use tokio::process::{ChildStdin, ChildStdout};
use tokio_util::codec::{FramedRead, LinesCodec};
use tokio_util::sync::CancellationToken;

pub const TESTED_VERSION: &str = "0.153.4";

pub fn version_warning(installed: &str) -> Option<String> {
    let expected = format!("codex-cli {TESTED_VERSION}");
    (installed.trim() != expected).then(|| {
        crate::runner::version_warning(
            Backend::Codex,
            installed,
            &format!("tested {expected}. Pin the tested CLI before live commissioning"),
        )
    })
}

pub struct Codex {
    _child: GroupChild,
    input: ChildStdin,
    output: FramedRead<ChildStdout, LinesCodec>,
    serial: u64,
    pending: VecDeque<(Value, usize)>,
    pending_bytes: usize,
    timeout: u64,
    cancel: CancellationToken,
    store: Store,
    entity: String,
}
impl Codex {
    pub async fn connect(
        config: &Config,
        cwd: &Path,
        store: Store,
        entity: &str,
        cancel: CancellationToken,
    ) -> Result<Self> {
        ensure!(!cancel.is_cancelled(), "Session cancelled");
        let mut child = GroupChild::new(
            process::command(&config.codex_binary, cwd)
                .args(["app-server", "--listen", "stdio://"])
                .stdin(Stdio::piped())
                .stderr(Stdio::null())
                .spawn()
                .context(
                    "Could not start Codex app-server; install and authenticate Codex on this host",
                )?,
        );
        let input = child.0.stdin.take().unwrap();
        let output = FramedRead::new(
            child.0.stdout.take().unwrap(),
            LinesCodec::new_with_max_length(MAX_MESSAGE),
        );
        let mut s = Self {
            _child: child,
            input,
            output,
            serial: 0,
            pending: VecDeque::new(),
            pending_bytes: 0,
            timeout: config.session_timeout_seconds,
            cancel,
            store,
            entity: entity.into(),
        };
        s.rpc("initialize",json!({"clientInfo":{"name":"octomus_agent","title":"Octomus Agent","version":env!("CARGO_PKG_VERSION")},"capabilities":{"experimentalApi":false}})).await?;
        s.send(json!({"method":"initialized","params":{}})).await?;
        Ok(s)
    }
    /// Authentication state plus the installed CLI's version against the tested
    /// baseline.
    pub async fn diagnostics(
        &mut self,
        config: &Config,
        cwd: &Path,
        cancel: &CancellationToken,
    ) -> Result<Value> {
        let account = self
            .rpc("account/read", json!({"refreshToken":false}))
            .await?;
        ensure!(
            account["requiresOpenaiAuth"] == false || !account["account"].is_null(),
            "Codex authentication is missing; run codex login as the service user"
        );
        let version = crate::process::run_machine(
            &config.codex_binary,
            &["--version"],
            cwd,
            config.command_timeout_seconds.min(60),
            cancel,
        )
        .await?
        .trim()
        .to_owned();
        Ok(crate::runner::diagnostics_value(
            Backend::Codex,
            &version,
            TESTED_VERSION,
            version_warning(&version),
        ))
    }
    async fn send(&mut self, value: Value) -> Result<()> {
        let payload = format!("{value}\n");
        let write = async {
            self.input.write_all(payload.as_bytes()).await?;
            self.input.flush().await
        };
        process::bounded(
            self.timeout.min(60),
            &self.cancel,
            "Codex write timed out",
            write,
        )
        .await??;
        Ok(())
    }
    /// Writes without checking `self.cancel`: cleanup requests must still reach the
    /// server after operator cancellation. Callers bound the wait themselves.
    async fn send_best_effort(&mut self, value: Value) -> Result<()> {
        let payload = format!("{value}\n");
        self.input.write_all(payload.as_bytes()).await?;
        self.input.flush().await?;
        Ok(())
    }
    async fn receive(&mut self) -> Result<Value> {
        let line = process::bounded(
            self.timeout,
            &self.cancel,
            "Codex response timed out",
            self.output.next(),
        )
        .await?
        .context("Codex app-server disconnected")??;
        let v: Value = serde_json::from_str(&line).context("Invalid app-server JSON")?;
        if v.get("method").is_some() && v.get("id").is_some() {
            self.send(json!({"id":v["id"],"error":{"code":-32000,"message":"Octomus unattended mode cannot answer interactive requests"}})).await?;
            bail!(
                "Codex requested interactive input ({}); task blocked",
                v["method"]
            );
        }
        Ok(v)
    }
    pub async fn rpc(&mut self, method: &str, params: Value) -> Result<Value> {
        self.serial += 1;
        let id = self.serial;
        self.send(json!({"id":id,"method":method,"params":params}))
            .await?;
        // A turn can emit notifications before the request response; preserve their order.
        let deadline = tokio::time::Instant::now() + Duration::from_secs(60);
        let cancel = self.cancel.clone();
        loop {
            let v = process::bounded_at(deadline, &cancel, "Codex RPC timed out", self.receive())
                .await??;
            if v["id"] == id {
                if let Some(e) = v.get("error") {
                    bail!("Codex {method}: {}", crate::store::redact(&e.to_string()));
                }
                return v.get("result").cloned().context("Missing RPC result");
            }
            ensure!(
                self.pending.len() < 10000,
                "Codex notification backlog exceeded"
            );
            let size = v.to_string().len();
            ensure!(
                self.pending_bytes + size <= 8 * 1024 * 1024,
                "Codex notification backlog exceeds 8 MB"
            );
            self.pending_bytes += size;
            self.pending.push_back((v, size));
        }
    }
    pub async fn models(&mut self) -> Result<Vec<Model>> {
        let mut out = vec![];
        let mut cursor = Value::Null;
        loop {
            let v = self
                .rpc(
                    "model/list",
                    json!({"limit":100,"cursor":cursor,"includeHidden":true}),
                )
                .await?;
            for m in v["data"].as_array().context("Invalid model catalog")? {
                out.push(Model {
                    backend: Backend::Codex,
                    provider: None,
                    provider_name: None,
                    model: m["model"]
                        .as_str()
                        .context("Invalid Codex model identity")?
                        .into(),
                    display_name: m["displayName"].as_str().unwrap_or("").into(),
                    efforts: m["supportedReasoningEfforts"]
                        .as_array()
                        .context("Invalid Codex reasoning catalog")?
                        .iter()
                        .map(|e| {
                            e["reasoningEffort"]
                                .as_str()
                                .map(str::to_owned)
                                .context("Invalid reasoning effort")
                        })
                        .collect::<Result<_>>()?,
                    variants: vec![],
                    available: true,
                    unavailable_reason: None,
                });
            }
            cursor = v["nextCursor"].clone();
            if cursor.is_null() {
                break;
            }
            ensure!(out.len() < 10000, "Invalid model pagination");
        }
        Ok(out)
    }
    pub async fn start(
        &mut self,
        route: &Route,
        cwd: &Path,
        resume: Option<&str>,
    ) -> Result<String> {
        route.require_backend(Backend::Codex)?;
        let mut params = json!({"model":route.model,"cwd":cwd,"approvalPolicy":"never","sandbox":"danger-full-access","config":{"model_reasoning_effort":route.effort},"developerInstructions":crate::runner::WORKER_INSTRUCTIONS});
        let method = if let Some(id) = resume {
            params["threadId"] = id.into();
            "thread/resume"
        } else {
            "thread/start"
        };
        let v = self.rpc(method, params).await?;
        ensure!(
            v["model"].as_str() == Some(&route.model),
            "Runtime substituted the requested model"
        );
        ensure!(
            v["reasoningEffort"].as_str() == Some(&route.effort),
            "Runtime substituted the requested reasoning effort"
        );
        ensure!(
            v["approvalPolicy"] == "never" && v["sandbox"]["type"] == "dangerFullAccess",
            "Runtime did not grant the configured no-sandbox/never-approve policy"
        );
        let identity = v["thread"]["id"]
            .as_str()
            .context("Missing Codex thread identity")?;
        uuid::Uuid::parse_str(identity).context("Invalid Codex thread identity")?;
        ensure!(
            resume.is_none_or(|id| identity == id),
            "Codex substituted the requested thread"
        );
        Ok(identity.to_owned())
    }
    pub async fn turn(
        &mut self,
        thread: &str,
        route: &Route,
        cwd: &Path,
        prompt: &str,
        schema: Option<Value>,
    ) -> Result<String> {
        route.require_backend(Backend::Codex)?;
        let mut params = json!({"threadId":thread,"cwd":cwd,"model":route.model,"effort":route.effort,"approvalPolicy":"never","sandboxPolicy":{"type":"dangerFullAccess"},"input":[{"type":"text","text":prompt,"text_elements":[]}]});
        if let Some(schema) = schema {
            params["outputSchema"] = schema;
        }
        let v = self.rpc("turn/start", params).await?;
        let turn = v["turn"]["id"]
            .as_str()
            .context("Missing turn identity")?
            .to_owned();
        let deadline = tokio::time::Instant::now() + Duration::from_secs(self.timeout);
        let cancel = self.cancel.clone();
        let result: Result<String> = async {
            let mut answer = String::new();
            loop {
                ensure!(!self.cancel.is_cancelled(), "Session cancelled");
                ensure!(
                    tokio::time::Instant::now() < deadline,
                    "Codex session time limit exceeded"
                );
                let event = if let Some((v, size)) = self.pending.pop_front() {
                    self.pending_bytes -= size;
                    v
                } else {
                    process::bounded_at(
                        deadline,
                        &cancel,
                        "Codex session time limit exceeded",
                        self.receive(),
                    )
                    .await??
                };
                if event["params"]["threadId"].as_str() != Some(thread) {
                    continue;
                }
                if event["params"]["turnId"]
                    .as_str()
                    .is_some_and(|id| id != turn)
                {
                    continue;
                }
                match event["method"].as_str().unwrap_or("") {
                    "item/completed" => {
                        let item = &event["params"]["item"];
                        if item["type"] == "agentMessage" && item["phase"] != "commentary" {
                            answer = item["text"].as_str().unwrap_or("").to_owned();
                        }
                        // Only record metadata, never raw tool arguments or command output from session notifications.
                        self.store.event(
                            &self.entity,
                            "session_progress",
                            &format!(
                                "{} · {} · {}",
                                thread,
                                item["type"].as_str().unwrap_or("item"),
                                item["status"].as_str().unwrap_or("completed")
                            ),
                        )?;
                    }
                    "turn/completed" if event["params"]["turn"]["id"] == turn => {
                        ensure!(
                            event["params"]["turn"]["status"] == "completed",
                            "Codex turn did not complete successfully: {}",
                            crate::store::redact(&event["params"]["turn"]["error"].to_string())
                        );
                        ensure!(!answer.trim().is_empty(), "Codex returned no final result");
                        break Ok(answer);
                    }
                    _ => {}
                }
            }
        }
        .await;
        if result.is_err() {
            // Best-effort interrupt bypasses the cancel select so it still reaches
            // the server after operator cancellation; the response is not awaited.
            self.serial += 1;
            let _ = tokio::time::timeout(
                Duration::from_secs(5),
                self.send_best_effort(
                    json!({"id":self.serial,"method":"turn/interrupt","params":{"threadId":thread,"turnId":turn}}),
                ),
            )
            .await;
        }
        result
    }
}

File: internal/runner/codex.go
package runner

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/google/uuid"

	octomus "github.com/tyk-swe/octomus-agent"
	"github.com/tyk-swe/octomus-agent/internal/config"
	"github.com/tyk-swe/octomus-agent/internal/process"
	"github.com/tyk-swe/octomus-agent/internal/schemas"
	"github.com/tyk-swe/octomus-agent/internal/store"
)

const CodexTestedVersion = "0.153.4"

func CodexVersionWarning(installed string) *string {
	expected := "codex-cli " + CodexTestedVersion
	if strings.TrimSpace(installed) == expected {
		return nil
	}
	warning := VersionWarning(config.BackendCodex, installed,
		fmt.Sprintf("tested %s. Pin the tested CLI before live commissioning", expected))
	return &warning
}

var errCodexDisconnected = errors.New("Codex app-server disconnected")

type queuedMessage struct {
	value map[string]any
	size  int
}

// backlog keeps, in order, the notifications that arrive before an RPC
// response, bounded to 10000 messages and 8 MiB of compact JSON.
type backlog struct {
	items []queuedMessage
	bytes int
}

func (b *backlog) push(value map[string]any) error {
	if len(b.items) >= 10000 {
		return fmt.Errorf("Codex notification backlog exceeded")
	}
	encoded, err := marshal(value)
	if err != nil {
		return err
	}
	if b.bytes+len(encoded) > 8*1024*1024 {
		return fmt.Errorf("Codex notification backlog exceeds 8 MB")
	}
	b.bytes += len(encoded)
	b.items = append(b.items, queuedMessage{value, len(encoded)})
	return nil
}

// pop removes and returns the oldest queued notification.
func (b *backlog) pop() (map[string]any, bool) {
	if len(b.items) == 0 {
		return nil, false
	}
	queued := b.items[0]
	b.items[0] = queuedMessage{}
	b.items = b.items[1:]
	if len(b.items) == 0 {
		b.items = nil
	}
	b.bytes -= queued.size
	return queued.value, true
}

// Codex owns a `codex app-server --listen stdio://` child and speaks its
// newline-delimited JSON-RPC protocol.
type Codex struct {
	child    *process.GroupChild
	stdin    *os.File
	stdout   *os.File
	lines    chan lineResult
	serial   uint64
	pending  backlog
	timeout  uint64
	ctx      context.Context
	state    *store.Store
	entity   string
	waitCh   chan error
	done     chan struct{}
	once     sync.Once
	closeErr error
	// binary and commandTimeout bound the Diagnostics version check.
	binary         string
	commandTimeout uint64
}

func ConnectCodex(ctx context.Context, cfg config.Config, cwd string, state *store.Store, entity string) (*Codex, error) {
	if ctx.Err() != nil {
		return nil, process.ErrSessionCancelled
	}
	cmd := process.Command(cfg.CodexBinary, cwd)
	cmd.Args = append(cmd.Args, "app-server", "--listen", "stdio://")
	stdinR, stdinW, err := os.Pipe()
	if err != nil {
		return nil, err
	}
	stdoutR, stdoutW, err := os.Pipe()
	if err != nil {
		stdinR.Close()
		stdinW.Close()
		return nil, err
	}
	cmd.Stdin = stdinR
	cmd.Stdout = stdoutW
	// Stderr is kept only as a bounded tail that explains a connect failure.
	tail := &stderrTail{}
	cmd.Stderr = tail
	cmd.WaitDelay = stderrWaitDelay
	if err := cmd.Start(); err != nil {
		stdinR.Close()
		stdinW.Close()
		stdoutR.Close()
		stdoutW.Close()
		return nil, fmt.Errorf("Could not start Codex app-server; install and authenticate Codex on this host: %w", err)
	}
	// The child holds its own pipe ends now; the parent's copies must close or
	// the reader would never see end-of-stream.
	stdinR.Close()
	stdoutW.Close()
	done := make(chan struct{})
	c := &Codex{
		child:   process.NewGroupChild(cmd),
		stdin:   stdinW,
		stdout:  stdoutR,
		lines:   lineReader(stdoutR, MaxMessage, done),
		timeout: cfg.SessionTimeoutSeconds,
		ctx:     ctx,
		state:   state,
		entity:  entity,
		waitCh:  make(chan error, 1),
		done:    done,

		binary:         cfg.CodexBinary,
		commandTimeout: cfg.CommandTimeoutSeconds,
	}
	go func() { c.waitCh <- cmd.Wait() }()
	fail := func(err error) (*Codex, error) {
		_ = c.Close()
		return nil, tail.explain(err)
	}
	if _, err := c.rpc("initialize", map[string]any{
		"clientInfo":   map[string]any{"name": "octomus_agent", "title": "Octomus Agent", "version": octomus.Version},
		"capabilities": map[string]any{"experimentalApi": false},
	}); err != nil {
		return fail(err)
	}
	if err := c.send(map[string]any{"method": "initialized", "params": map[string]any{}}); err != nil {
		return fail(err)
	}
	return c, nil
}

// Diagnostics reports authentication state plus the installed CLI's version
// against the tested baseline.
func (c *Codex) Diagnostics(cwd string) (map[string]any, error) {
	account, err := c.rpc("account/read", map[string]any{"refreshToken": false})
	if err != nil {
		return nil, err
	}
	m, _ := asObject(account)
	if m["requiresOpenaiAuth"] != false && m["account"] == nil {
		return nil, fmt.Errorf("Codex authentication is missing; run codex login as the service user")
	}
	version, err := process.RunMachine(c.ctx, c.binary, []string{"--version"}, cwd, min(c.commandTimeout, 60))
	if err != nil {
		return nil, err
	}
	version = strings.TrimSpace(version)
	return diagnosticsValue(config.BackendCodex, version, CodexTestedVersion, CodexVersionWarning(version)), nil
}

// framed marshals a protocol message and applies the exact outbound bound to
// the JSON content: MaxMessage bytes of content is accepted, one byte more is
// rejected, and the newline terminator is not counted against the bound.
func framed(value map[string]any) (string, error) {
	payload, err := marshal(value)
	if err != nil {
		return "", err
	}
	if len(payload) > MaxMessage {
		return "", errors.New("Codex message exceeds the 16 MB protocol limit")
	}
	return payload + "\n", nil
}

func (c *Codex) send(value map[string]any) error {
	payload, err := framed(value)
	if err != nil {
		return err
	}
	_, err = process.Bounded(c.ctx, min(c.timeout, 60), "Codex write timed out", func(wctx context.Context) (struct{}, error) {
		return struct{}{}, writeAll(wctx, c.stdin, []byte(payload))
	})
	return err
}

// sendBestEffort writes without checking the owner context: cleanup requests
// must still reach the server after operator cancellation. The caller's bound
// supplies the write context.
func (c *Codex) sendBestEffort(ctx context.Context, value map[string]any) error {
	payload, err := framed(value)
	if err != nil {
		return err
	}
	return writeAll(ctx, c.stdin, []byte(payload))
}

// writeAll observes ctx so a write blocked on a full pipe unwinds with the
// caller instead of stranding a goroutine. A rolling write deadline of at
// most 250ms, clipped to the context deadline, turns a blocking write into a
// retryable timeout while the context is live.
func writeAll(ctx context.Context, w *os.File, data []byte) error {
	defer w.SetWriteDeadline(time.Time{})
	for len(data) > 0 {
		if err := ctx.Err(); err != nil {
			return err
		}
		deadline := time.Now().Add(250 * time.Millisecond)
		if limit, ok := ctx.Deadline(); ok && limit.Before(deadline) {
			deadline = limit
		}
		if err := w.SetWriteDeadline(deadline); err != nil {
			return err
		}
		n, err := w.Write(data)
		if n > 0 {
			data = data[n:]
		}
		if err != nil {
			if errors.Is(err, os.ErrDeadlineExceeded) {
				if ctx.Err() != nil {
					return ctx.Err()
				}
				continue
			}
			return err
		}
	}
	return nil
}

// receive reads one protocol message by deadline, answering and rejecting
// interactive JSON-RPC requests. what names the bound in the timeout error.
func (c *Codex) receive(deadline time.Time, what string) (map[string]any, error) {
	timer := time.NewTimer(time.Until(deadline))
	defer timer.Stop()
	var r lineResult
	select {
	case line, ok := <-c.lines:
		if !ok {
			return nil, errCodexDisconnected
		}
		r = line
	case <-timer.C:
		return nil, fmt.Errorf("%s: deadline has elapsed", what)
	case <-c.ctx.Done():
		return nil, process.ErrSessionCancelled
	}
	if r.err != nil {
		return nil, r.err
	}
	decoded, err := decodeJSON(r.line)
	if err != nil {
		return nil, fmt.Errorf("Invalid app-server JSON: %w", err)
	}
	v, _ := asObject(decoded)
	if v == nil {
		return map[string]any{}, nil
	}
	id, hasID := v["id"]
	if _, hasMethod := v["method"]; hasMethod && hasID {
		if err := c.send(map[string]any{
			"id":    id,
			"error": map[string]any{"code": -32000, "message": "Octomus unattended mode cannot answer interactive requests"},
		}); err != nil {
			return nil, err
		}
		name, _ := marshal(v["method"])
		return nil, fmt.Errorf("Codex requested interactive input (%s); task blocked", name)
	}
	return v, nil
}

func (c *Codex) rpc(method string, params map[string]any) (any, error) {
	c.serial++
	id := c.serial
	if err := c.send(map[string]any{"id": id, "method": method, "params": params}); err != nil {
		return nil, err
	}
	// A turn can emit notifications before the request response; preserve
	// their order. The whole call is bounded at 60 s and each message by the
	// session timeout, whichever ends first.
	deadline := time.Now().Add(60 * time.Second)
	for {
		bound, what := deadline, "Codex RPC timed out"
		if perMessage := time.Now().Add(time.Duration(c.timeout) * time.Second); perMessage.Before(bound) {
			bound, what = perMessage, "Codex response timed out"
		}
		v, err := c.receive(bound, what)
		if err != nil {
			return nil, err
		}
		if idMatches(v["id"], id) {
			if e, hasErr := v["error"]; hasErr {
				encoded, _ := marshal(e)
				return nil, fmt.Errorf("Codex %s: %s", method, store.Redact(encoded))
			}
			result, ok := v["result"]
			if !ok {
				return nil, fmt.Errorf("Missing RPC result")
			}
			return result, nil
		}
		if err := c.pending.push(v); err != nil {
			return nil, err
		}
	}
}

func idMatches(v any, id uint64) bool {
	n, ok := v.(json.Number)
	return ok && n.String() == strconv.FormatUint(id, 10)
}

// Models lists the catalog with model/list pagination.
func (c *Codex) Models(cwd string) ([]Model, error) {
	out := []Model{}
	var cursor any
	for {
		v, err := c.rpc("model/list", map[string]any{"limit": 100, "cursor": cursor, "includeHidden": true})
		if err != nil {
			return nil, err
		}
		result, _ := asObject(v)
		data, ok := asArray(result["data"])
		if !ok {
			return nil, fmt.Errorf("Invalid model catalog")
		}
		for _, item := range data {
			m, _ := asObject(item)
			name, ok := strAt(m, "model")
			if !ok {
				return nil, fmt.Errorf("Invalid Codex model identity")
			}
			display, _ := strAt(m, "displayName")
			effortList, ok := asArray(m["supportedReasoningEfforts"])
			if !ok {
				return nil, fmt.Errorf("Invalid Codex reasoning catalog")
			}
			efforts := []string{}
			for _, e := range effortList {
				entry, _ := asObject(e)
				effort, ok := strAt(entry, "reasoningEffort")
				if !ok {
					return nil, fmt.Errorf("Invalid reasoning effort")
				}
				efforts = append(efforts, effort)
			}
			out = append(out, Model{
				Backend:     config.BackendCodex,
				Model:       name,
				DisplayName: display,
				Efforts:     efforts,
				Variants:    []string{},
				Available:   true,
			})
		}
		cursor = result["nextCursor"]
		if cursor == nil {
			break
		}
		// A continuing page must make progress: an empty page with a cursor,
		// or a catalog past the model bound, would page forever.
		if len(data) == 0 || len(out) >= 10000 {
			return nil, fmt.Errorf("Invalid model pagination")
		}
	}
	return out, nil
}

func (c *Codex) Start(route config.Route, cwd string, resume *string) (string, error) {
	if err := requireRoute(route, config.BackendCodex); err != nil {
		return "", err
	}
	params := map[string]any{
		"model":                 route.Model,
		"cwd":                   cwd,
		"approvalPolicy":        "never",
		"sandbox":               "danger-full-access",
		"config":                map[string]any{"model_reasoning_effort": route.Effort},
		"developerInstructions": WorkerInstructions,
	}
	method := "thread/start"
	if resume != nil {
		params["threadId"] = *resume
		method = "thread/resume"
	}
	v, err := c.rpc(method, params)
	if err != nil {
		return "", err
	}
	result, _ := asObject(v)
	if s, _ := strAt(result, "model"); s != route.Model {
		return "", fmt.Errorf("Runtime substituted the requested model")
	}
	if s, _ := strAt(result, "reasoningEffort"); s != route.Effort {
		return "", fmt.Errorf("Runtime substituted the requested reasoning effort")
	}
	sandbox, _ := asObject(result["sandbox"])
	sandboxType, _ := strAt(sandbox, "type")
	if result["approvalPolicy"] != "never" || sandboxType != "dangerFullAccess" {
		return "", fmt.Errorf("Runtime did not grant the configured no-sandbox/never-approve policy")
	}
	thread, _ := asObject(result["thread"])
	identity, ok := strAt(thread, "id")
	if !ok {
		return "", fmt.Errorf("Missing Codex thread identity")
	}
	if _, err := uuid.Parse(identity); err != nil {
		return "", fmt.Errorf("Invalid Codex thread identity: %w", err)
	}
	if resume != nil && identity != *resume {
		return "", fmt.Errorf("Codex substituted the requested thread")
	}
	return identity, nil
}

func (c *Codex) Turn(session string, route config.Route, cwd, prompt string, schema schemas.Schema) (string, error) {
	if err := requireRoute(route, config.BackendCodex); err != nil {
		return "", err
	}
	answer, err := c.turn(session, route, cwd, prompt, schema)
	if err != nil {
		return "", err
	}
	return FinishTurn(answer, schema)
}

// turn starts a turn and follows it to completion; any failure after the
// turn has started interrupts it.
func (c *Codex) turn(thread string, route config.Route, cwd, prompt string, schema schemas.Schema) (string, error) {
	turn, err := c.startTurn(thread, route, cwd, prompt, schema)
	if err != nil {
		return "", err
	}
	deadline := time.Now().Add(time.Duration(c.timeout) * time.Second)
	answer, err := c.awaitTurn(thread, turn, deadline)
	if err != nil {
		c.interrupt(thread, turn)
		return "", err
	}
	return answer, nil
}

// startTurn submits turn/start and returns the new turn's identity.
func (c *Codex) startTurn(thread string, route config.Route, cwd, prompt string, schema schemas.Schema) (string, error) {
	params := map[string]any{
		"threadId":       thread,
		"cwd":            cwd,
		"model":          route.Model,
		"effort":         route.Effort,
		"approvalPolicy": "never",
		"sandboxPolicy":  map[string]any{"type": "dangerFullAccess"},
		"input":          []any{map[string]any{"type": "text", "text": prompt, "text_elements": []any{}}},
	}
	if schema != nil {
		params["outputSchema"] = schema
	}
	v, err := c.rpc("turn/start", params)
	if err != nil {
		return "", err
	}
	result, _ := asObject(v)
	turnObj, _ := asObject(result["turn"])
	turn, ok := strAt(turnObj, "id")
	if !ok {
		return "", fmt.Errorf("Missing turn identity")
	}
	return turn, nil
}

// nextEvent checks cancellation and the turn deadline, then returns the
// oldest queued notification or receives the next message.
func (c *Codex) nextEvent(deadline time.Time) (map[string]any, error) {
	if c.ctx.Err() != nil {
		return nil, process.ErrSessionCancelled
	}
	if !time.Now().Before(deadline) {
		return nil, fmt.Errorf("Codex session time limit exceeded")
	}
	if event, ok := c.pending.pop(); ok {
		return event, nil
	}
	return c.receive(deadline, "Codex session time limit exceeded")
}

// awaitTurn follows the turn's events until it completes and returns the
// final agent message.
func (c *Codex) awaitTurn(thread, turn string, deadline time.Time) (string, error) {
	var answer string
	for {
		event, err := c.nextEvent(deadline)
		if err != nil {
			return "", err
		}
		params, _ := asObject(event["params"])
		if s, _ := strAt(params, "threadId"); s != thread {
			continue
		}
		if id, ok := strAt(params, "turnId"); ok && id != turn {
			continue
		}
		method, _ := strAt(event, "method")
		switch method {
		case "item/completed":
			item, _ := asObject(params["item"])
			itemType, _ := strAt(item, "type")
			phase, _ := strAt(item, "phase")
			if itemType == "agentMessage" && phase != "commentary" {
				answer, _ = strAt(item, "text")
			}
			// Only record metadata, never raw tool arguments or command
			// output from session notifications.
			itemStatus, _ := strAt(item, "status")
			if itemType == "" {
				itemType = "item"
			}
			if itemStatus == "" {
				itemStatus = "completed"
			}
			if err := c.state.Event(c.entity, "session_progress", fmt.Sprintf("%s · %s · %s", thread, itemType, itemStatus)); err != nil {
				return "", err
			}
		case "turn/completed":
			completed, _ := asObject(params["turn"])
			if s, _ := strAt(completed, "id"); s != turn {
				continue
			}
			if s, _ := strAt(completed, "status"); s != "completed" {
				encoded, _ := marshal(completed["error"])
				return "", fmt.Errorf("Codex turn did not complete successfully: %s", store.Redact(encoded))
			}
			if strings.TrimSpace(answer) == "" {
				return "", fmt.Errorf("Codex returned no final result")
			}
			return answer, nil
		}
	}
}

// interrupt sends a best-effort turn/interrupt that bypasses the owner
// context, so it still reaches the server after operator cancellation; the
// response is not awaited.
func (c *Codex) interrupt(thread, turn string) {
	c.serial++
	id := c.serial
	_, _ = process.Bounded(context.Background(), 5, "Codex interrupt timed out", func(wctx context.Context) (struct{}, error) {
		return struct{}{}, c.sendBestEffort(wctx, map[string]any{
			"id":     id,
			"method": "turn/interrupt",
			"params": map[string]any{"threadId": thread, "turnId": turn},
		})
	})
}

// Close kills the process group, closes the pipes, and joins both the line
// reader and the child wait with a bounded cleanup. Closing done makes the
// reader abandon pending sends so it always terminates. Idempotent.
func (c *Codex) Close() error {
	c.once.Do(func() {
		close(c.done)
		c.child.Close()
		c.stdin.Close()
		c.stdout.Close()
		c.closeErr = joinOwned(c.waitCh, drained(c.lines), "Codex app-server did not exit during cleanup")
	})
	return c.closeErr
}
